hbase-oss/src/main/java/org/apache/hadoop/hbase/oss/sync/TreeLockManager.java

```diff
@@ -0,0 +1,498 @@
+   *
+   * @param path Path to lock
+   */
+  protected void treeReadLock(Path p) throws IOException {
+    int outerRetries = 0;
+    do {
+      int innerRetries = 0;
+      do {
+        // If there's a write lock above us, wait
+        if (writeLockAbove(p)) {
+          LOG.warn("Blocked waiting for some parent write lock, waiting: {}",
+                p);
+          continue;
+        }
+        break;
+      } while (retryBackoff(innerRetries++));
+      // Try obtain the read-lock just for our node
+      readLock(p);
+      // If there's a write lock above us, release the lock and try again
+      if (writeLockAbove(p)) {
+        LOG.warn("Blocked waiting for some parent write lock, retrying: {}",
+              p);
+        readUnlock(p);
+        continue;
+      }
+      break;
+    } while (retryBackoff(outerRetries++));
+  }
+
+  /**
+   * Acquires an exclusive lock on a single path to create or append to a file.
+   * This is required for createNonRecursive() as well as other operations to be
+   * atomic because the underlying file may not be created until all data has
+   * been written.
+   *
+   * @param path Path of the create operation
+   * @return AutoLock to release this path
+   */
+  public AutoLock lockWrite(Path rawPath) throws IOException {
+    Path path = norm(rawPath);
+    LOG.debug("About to lock for create / write: {}", rawPath);
+    treeWriteLock(path);
+    return new AutoLock() {
+      public void close() throws IOException {
+        LOG.debug("About to unlock after create / write: {}", path);
+        writeUnlock(path);
+      }
+    };
+  }
+
+  /**
+   * Acquires an exclusive lock on a single path and then cleans up the lock
+   * and those of all children. The lock ensures this doesn't interfere with any
+   * renames or other listing operations above this path.
+   *
+   * @param path Path of the create operation
+   * @return AutoLock to release this path
+   */
+  public AutoLock lockDelete(Path rawPath) throws IOException {
+    Path path = norm(rawPath);
+    LOG.debug("About to lock for delete: {}", path);
+    treeWriteLock(path);
+    return new AutoLock() {
+      public void close() throws IOException {
+        LOG.debug("About to recursively delete locks: {}", path);
+        recursiveDelete(path);
+      }
+    };
+  }
+
+  /**
+   * Acquires a lock on a single path to run a listing operation. We need to
+   * ensure that the listing is not generated mid-rename. This will lock all
+   * children of the root path as well, which is only necessary for recursive
+   * listings. Other listings should only need a read lock on the root and all
+   * children, but that is not implemented.
+   *
+   * @param path Root of the listing operation
+   * @return AutoCloseable to release this path
+   */
+  public AutoLock lockListing(Path rawPath) throws IOException {
+    Path path = norm(rawPath);
+    LOG.debug("About to lock for listing: {}", path);
+    treeWriteLock(path);
```

Comment:
Why do we need treeWriteLock when listing? Just found in my tests that trying restart HMaster while RSes are performing puts causes Master to never complete initialization. This is because during startup, HMaster tries to list tables directories, by doing a FileSystem.globStatus call on path "/s3a/wellington-hbase-s3/hbase/data", for example. With RSes busy writing hfiles for related table regions, Master ends up never acquiring write lock for "/s3a/wellington-hbase-s3/hbase/data", in this case. I had tested replace the write lock for a read one, this seems to sort the issue, tests are looking promising so far. But ain't sure on @mackrorysd motivations for defining a write lock for listing.

Reply:
Although listing is a read operation, the key is that it needs an *exclusive* lock, otherwise it's vulnerable to include partial results of renames that are currently occurring in a child directory. One alternative I had considered is for non-recursive listings, getting the results outside of a lock, read-locking the individual children, re-doing the listing with those held, and returning if nothing has changed since obtaining the lock. If it's changed, retry. It's a little convoluted but it's safe. Other ideas welcome - but I'm hesitant to relax the constraints for listings in general.

Reply:
Right, maybe we can relax those for hbase? This specific case wouldn't be an issue, as it's not a recursive list, and that directory level is only changed by master anyway, which is still starting at the point.

Reply:
Oh yeah we can absolutely relax those for HBase. Remember, 2 of the key design principles here are (a) start strict and relax as we know we need to, and (b) keep it separate from S3Guard so we can relax on the HBase community's terms alone. If no other listing needs to ensure it doesn't see in-progress renames, or if we can distinguish renames that do from those that don't, we can definitely relax the locking and just make sure it's well documented how-and-why.

Reply:
Got it. I'm going ahead with testing that simple listing approach for now. Will keep you updated on the results.

Reply:
So I'm not sure we can globally make that distinction. Looking at the use of listStatus in IncrementalBackupManager, etc., it certainly seems to me that it needs to ensure it doesn't see inprogress renames, etc. Unless there's another mechanism preventing it from seeing in-progress compactions, etc.

Reply:
This would bring us to the point of requiring some distinct code paths in HBase, but what if we added a "listStatusLockless" for cases where we know it's a problem but we know it's safe? Of course if this is going through globStatus, there's a lot of Globbing code that would ultimately call the regular listStatus. So maybe we need a setting to make listings only do readLock JVM-wide if we know the master can do that.

Reply:
Better idea: I'm just going to modify treeWriteLock and associated functions to take a "depth" argument. Most listings only require looking 1 layer down. Otherwise the logic is unchanged: there's no need for a 2-phase listing where we make sure the results haven't changed after obtaining the lock, etc. Listings that do require a recursive lock still get one. There's still some weird cases like a delete lock ensuring that no one has an exclusive lock above it. But again, let's optimize that if it comes up. Testing a patch for this now and will share it when it's ready.